File: groupanagrams.js
//Objective is to group the anagrams together in a 1-D array.

let strings = ['eat', 'tea', 'tan', 'ate', 'nat', 'bat']


//O(n*klogk) where n is the length of the input array and k is the max length
//of a string in the input array.

let map = {}
let result = []

//For each string, sort it and map it to the hashmap 
for (let string of strings) {
    let temp = string 
    let sorted = string.split('').sort().join('')

    if (map[sorted] == undefined) {
        map[sorted] = [temp]
    } else {
        map[sorted].push(temp)
    }
}

for (let array in map) {
    result.push(map[array])
}

return result

File: groupanagrams.java
//Java Solution

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0) {
            return new ArrayList();
        }
        
        Map<String, List> result = new HashMap<String, List>();
        for (String str : strs) {
            char[] ca = str.toCharArray();
            Arrays.sort(ca);
            String sorted = String.valueOf(ca);
            
            if (!result.containsKey(sorted)) {
                result.put(sorted, new ArrayList());
            }
            result.get(sorted).add(str);
        }
        
        return new ArrayList(result.values());
    }
}